Investigate why no functions get inlined on binary builds

Split out of #13889 

```
$ git diff
diff --git a/Makefile.defs b/Makefile.defs
index 1eb900c7d191..011c5f7dbbc1 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -146,7 +146,7 @@ GO_BUILD_FLAGS += -ldflags '$(GO_BUILD_LDFLAGS) $(EXTRA_GO_BUILD_LDFLAGS)' -tags
 GO_TEST_FLAGS += -tags=$(call join-with-comma,$(GO_TAGS_FLAGS))
 
 ifeq ($(NOSTRIP),1)
-       GO_BUILD_FLAGS += -gcflags="all=-N -l"
+       GO_BUILD_FLAGS += -gcflags="all=-N -l -m"
 endif
 
 GO_BUILD += $(GO_BUILD_FLAGS)
$ cd daemon
$ make clean
$ make NOSTRIP=1 2>&1 | tee build.log
$ grep -E "(can inline|inlining call)" build.log
$ go tool objdump cilium-agent | grep -E "(rotl|fmix)64"
  murmur3.go:43		0x2707fd9		e862070000			CALL github.com/cilium/cilium/pkg/murmur3.rotl64(SB)	
  murmur3.go:46		0x2708011		e82a070000			CALL github.com/cilium/cilium/pkg/murmur3.rotl64(SB)	
  murmur3.go:50		0x2708060		e8db060000			CALL github.com/cilium/cilium/pkg/murmur3.rotl64(SB)	
  murmur3.go:53		0x2708098		e8a3060000			CALL github.com/cilium/cilium/pkg/murmur3.rotl64(SB)	
  murmur3.go:113	0x27081c0		e87b050000			CALL github.com/cilium/cilium/pkg/murmur3.rotl64(SB)	
  murmur3.go:122	0x270824a		e871050000			CALL github.com/cilium/cilium/pkg/murmur3.fmix64(SB)	
  murmur3.go:123	0x2708262		e859050000			CALL github.com/cilium/cilium/pkg/murmur3.fmix64(SB)	
  murmur3.go:85		0x27084c0		e87b020000			CALL github.com/cilium/cilium/pkg/murmur3.rotl64(SB)	
TEXT github.com/cilium/cilium/pkg/murmur3.rotl64(SB) /home/tklauser/go/src/github.com/cilium/cilium/pkg/murmur3/murmur3.go
TEXT github.com/cilium/cilium/pkg/murmur3.fmix64(SB) /home/tklauser/go/src/github.com/cilium/cilium/pkg/murmur3/murmur3.go
$
```

however, if we build a single package - in this case `pkg/murmur3` we get several inlined functions:

```
$ rm -rf $GOPATH/pkg/$(go env GOOS)_$(go env GOARCH)/github.com/cilium/cilium
$ go clean -cache && go clean -modcache
$ go install -gcflags=-m ./pkg/murmur3/
# github.com/cilium/cilium/pkg/murmur3
pkg/murmur3/murmur3.go:130:6: can inline rotl64
pkg/murmur3/murmur3.go:134:6: can inline fmix64
pkg/murmur3/murmur3.go:43:14: inlining call to rotl64
pkg/murmur3/murmur3.go:46:14: inlining call to rotl64
pkg/murmur3/murmur3.go:50:14: inlining call to rotl64
pkg/murmur3/murmur3.go:53:14: inlining call to rotl64
pkg/murmur3/murmur3.go:85:14: inlining call to rotl64
pkg/murmur3/murmur3.go:113:14: inlining call to rotl64
pkg/murmur3/murmur3.go:122:13: inlining call to fmix64
pkg/murmur3/murmur3.go:123:13: inlining call to fmix64
/tmp/go-build069986400/b001/_cgo_gotypes.go:14:6: can inline _Cgo_ptr
/tmp/go-build069986400/b001/_cgo_gotypes.go:49:6: can inline _Cfunc_CBytes
pkg/murmur3/murmur3_reference.go:154:89: inlining call to _Cfunc_CBytes
pkg/murmur3/murmur3.go:26:14: data does not escape
/tmp/go-build069986400/b001/_cgo_gotypes.go:14:15: leaking param: ptr to result ~r1 level=0
/tmp/go-build069986400/b001/_cgo_gotypes.go:37:43: assuming arg#2 is unsafe uintptr
/tmp/go-build069986400/b001/_cgo_gotypes.go:40:63: assuming arg#3 is unsafe uintptr
/tmp/go-build069986400/b001/_cgo_gotypes.go:40:72: assuming arg#4 is unsafe uintptr
/tmp/go-build069986400/b001/_cgo_gotypes.go:49:20: b does not escape
/tmp/go-build069986400/b001/_cgo_gotypes.go:61:33: leaking param: p0
/tmp/go-build069986400/b001/_cgo_gotypes.go:61:87: leaking param: p3
/tmp/go-build069986400/b001/_cgo_gotypes.go:65:11: p1 escapes to heap
/tmp/go-build069986400/b001/_cgo_gotypes.go:66:11: p2 escapes to heap
pkg/murmur3/murmur3_reference.go:152:17: leaking param: data
pkg/murmur3/murmur3_reference.go:153:13: make([]uint64, 2) escapes to heap
pkg/murmur3/murmur3_reference.go:154:2: func literal does not escape
pkg/murmur3/murmur3_reference.go:154:20: func literal does not escape
pkg/murmur3/murmur3_reference.go:155:3: *_cgoIndex3 escapes to heap
pkg/murmur3/murmur3_reference.go:154:55: _cgo0 escapes to heap
$ go tool objdump $GOPATH/pkg/$(go env GOOS)_$(go env GOARCH)/github.com/cilium/cilium/pkg/murmur3.a | grep -E "(rotl|fmix)64"
TEXT github.com/cilium/cilium/pkg/murmur3.rotl64(SB) gofile../home/tklauser/go/src/github.com/cilium/cilium/pkg/murmur3/murmur3.go
TEXT github.com/cilium/cilium/pkg/murmur3.fmix64(SB) gofile../home/tklauser/go/src/github.com/cilium/cilium/pkg/murmur3/murmur3.go
```

--- a/Makefile
+++ b/Makefile
@@ -4,6 +4,7 @@
 all: precheck build postcheck
 	@echo "Build finished."
 
+debug: export NOOPT=1
 debug: export NOSTRIP=1
 debug: all
 

--- a/Makefile.defs
+++ b/Makefile.defs
@@ -145,8 +145,8 @@ endif
 GO_BUILD_FLAGS += -ldflags '$(GO_BUILD_LDFLAGS) $(EXTRA_GO_BUILD_LDFLAGS)' -tags=$(call join-with-comma,$(GO_TAGS_FLAGS)) $(EXTRA_GO_BUILD_FLAGS)
 GO_TEST_FLAGS += -tags=$(call join-with-comma,$(GO_TAGS_FLAGS))
 
-ifeq ($(NOSTRIP),1)
-	GO_BUILD_FLAGS += -gcflags="all=-N -l"
+ifeq ($(NOOPT),1)
+    GO_BUILD_FLAGS += -gcflags="all=-N -l"
 endif
 
 GO_BUILD += $(GO_BUILD_FLAGS)

